constraints/miden-vm/rpo.air

```diff
@@ -0,0 +1,81 @@
+    ark_15: [14233283787297595392, 373995945117666496, 12489737547229155328, 17208448209698889728, 7139138592091307008, 13886063479078012928, 1664893052631119104, 0]
+    ark_16: [13792579614346651648, 1235734395091296000, 9500452585969031168, 8739495587021565952, 9012006439959783424, 14358505101923203072, 7629576092524553216, 0]
+    ark_17: [11614812331536766976, 14172757457833930752, 2054001340201038848, 17000774922218162176, 14619614108529063936, 7744142531772273664, 3485239601103661568, 0]
+    ark_18: [14871063686742261760, 707573103686350208, 12420704059284934656, 13533282547195531264, 1394813199588124416, 16135070735728404480, 9755891797164034048, 0]
+    ark_19: [10148237148793042944, 15453217512188186624, 355990932618543744, 525402848358706240, 4635111139507788800, 12290902521256030208, 15218148195153268736, 0]
+    ark_20: [4457428952329675776, 219777875004506016, 9071225051243524096, 16987541523062161408, 16217473952264204288, 12059913662657710080, 16460604813734957056, 0]
+    ark_21: [15590786458219171840, 17876696346199468032, 12766199826003447808, 5466806524462796800, 10782018226466330624, 16456018495793752064, 9643968136937730048, 0]
+    ark_22: [10063319113072093184, 17731621626449383424, 9045979173463557120, 14512769585918244864, 6844229992533661696, 4571485474751953408, 3611348709641382912, 0]
+    ark_23: [14200078843431360512, 2897136237748376064, 12934431667190679552, 10973956031244050432, 7446486531695178752, 17200392109565784064, 18256379591337758720, 0]
+
+### Helper functions ##############################################################################
+
+fn apply_mds(state: vector[12]) -> vector[12]:
+    return [sum([state[i] * mds_row[i] for i in 0..12]) for mds_row in MDS]
+
+fn apply_sbox(state: vector[12]) -> vector[12]:
+    return [s^7 for s in state]
+
+### RPO Air Constraints ##################################################################
+
+ev enforce_rpo_round([h[12]]):
+    let ark = [ark_0, ark_1, ark_2, ark_3, ark_4, ark_5, ark_6, ark_7, ark_8, ark_9, ark_10,
+        ark_11, ark_12, ark_13, ark_14, ark_15, ark_16, ark_17, ark_18, ark_19, ark_20, ark_21,
+        ark_22, ark_23]
+
+    # compute the state that should result from applying the first 5 operations of the RPO round to
+    # the current hash state.
+    let step1_initial = apply_mds(h)
+
+    # add constants
+    let step1_with_constants = [step1_initial[i] + ark[i] for i in 0..12]
+    
+    # apply sbox
+    let step1_with_sbox = apply_sbox(step1_with_constants)
+    
+    # apply mds
+    let step1_with_mds = apply_mds(step1_with_sbox)
+    
+    # add constants
+    let step1 = [step1_with_mds[i] + ark[i + 12] for i in 0..12]
+
+    # compute the state that should result from applying the inverse of the last operation of the
+    # RPO round to the next step of the computation.
+    let step2 = apply_sbox(h)
```

Comment:
I believe the S-box needs to be applied to the next state, rather than the current state. So, this should be something like:
```
let step2 = apply_sbox(h')
```
But I'm not sure we support this syntax (and not sure we should). Alternatively, we could change this to:
```
let step2 = [s'^7 for s in h]
```
And then we could probably git rid of `apply_sbox` function and replace line 68 with:
```
let step1_with_sbox = [s^7 for s in step1_with_constants]
```

Reply:
Thanks, fixed.

Reply:
```
let step2 = apply_sbox(h')
```

@bobbinth with the recent refactor we can support this, and we have used it in other places. It is because the argument passed to the function call is now a `SymbolAccess`, which includes an offset specification. Why do you not think we should? I actually find this clearer to read.

Reply:
```
let step2 = apply_sbox(h')
```

@bobbinth with the recent refactor we can support this, and we have used it in other places. It is because the argument passed to the function call is now a `SymbolAccess`, which includes an offset specification. Why do you not think we should? I actually find this clearer to read.

Reply:
My thinking was that it might introduce quite a bit of complexity - but if this is not too difficult to support, then it is fine.